pkg/providers/cloudstack/validator.go

```diff
@@ -48,83 +55,122 @@ type ProviderCmkClient interface {
-		return fmt.Errorf("CloudStackDatacenterConfig managementApiEndpoint is not set or is empty")
-	}
-	_, err := getHostnameFromUrl(datacenterConfig.Spec.ManagementApiEndpoint)
-	if err != nil {
-		return fmt.Errorf("checking management api endpoint: %v", err)
-	}
-	execConfig, err := decoder.ParseCloudStackSecret()
-	if err != nil {
-		return fmt.Errorf("parsing cloudstack secret: %v", err)
-	}
-	if execConfig.ManagementUrl != datacenterConfig.Spec.ManagementApiEndpoint {
-		return fmt.Errorf("cloudstack secret management url (%s) differs from cluster spec management url (%s)",
-			execConfig.ManagementUrl, datacenterConfig.Spec.ManagementApiEndpoint)
-	}
-
-	if err := v.validateDomainAndAccount(ctx, datacenterConfig); err != nil {
+	if err := v.generateLocalAvailabilityZones(ctx, datacenterConfig); err != nil {
 		return err
 	}
 
-	zones, errZone := v.cmk.ValidateZonesPresent(ctx, datacenterConfig.Spec.Zones)
-	if errZone != nil {
-		return fmt.Errorf("checking zones %v", errZone)
-	}
+	for _, az := range v.availabilityZones {
+		_, err := getHostnameFromUrl(az.ManagementApiEndpoint)
+		if err != nil {
+			return fmt.Errorf("checking management api endpoint: %v", err)
+		}
 
-	for _, zone := range datacenterConfig.Spec.Zones {
-		if len(zone.Network.Id) == 0 && len(zone.Network.Name) == 0 {
-			return fmt.Errorf("zone network is not set or is empty")
+		cmk := v.cmks[az.Name]
+		if cmk.GetManagementApiEndpoint(ctx) != az.ManagementApiEndpoint {
+			return fmt.Errorf("cloudstack secret management url (%s) differs from cluster spec management url (%s)",
+				cmk.GetManagementApiEndpoint(ctx), az.ManagementApiEndpoint)
 		}
-		err := v.cmk.ValidateNetworkPresent(ctx, domainId, zone, zones, datacenterConfig.Spec.Account, len(zones) > 1)
+
+		zones, err := cmk.ValidateZonesPresent(ctx, []anywherev1.CloudStackZone{az.CloudStackAvailabilityZone.Zone})
 		if err != nil {
-			return fmt.Errorf("checking network %v", err)
+			return err
+		}
+		if len(zones) > 0 {
+			az.CloudStackAvailabilityZone.Zone.Id = zones[0].Id
+		}
+		if len(az.CloudStackAvailabilityZone.Zone.Network.Id) == 0 && len(az.CloudStackAvailabilityZone.Zone.Network.Name) == 0 {
+			return fmt.Errorf("zone network is not set or is empty")
+		}
+		if err := cmk.ValidateNetworkPresent(ctx, az.DomainId, az.CloudStackAvailabilityZone.Zone, zones, az.Account, true); err != nil {
+			return err
 		}
 	}
 
 	logger.MarkPass("Datacenter validated")
 	return nil
 }
 
-func (v *Validator) validateDomainAndAccount(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
-	if (datacenterConfig.Spec.Domain != "" && datacenterConfig.Spec.Account == "") ||
-		(datacenterConfig.Spec.Domain == "" && datacenterConfig.Spec.Account != "") {
-		return fmt.Errorf("both domain and account must be specified or none of them must be specified")
+func (v *Validator) generateLocalAvailabilityZones(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
```

Comment:
If we convert to the new schema using the `SetDefaults` method, we won't need this function. Then we'll know that going into the validator, we only have to check the new fields, not the old ones. Although, once you change to using the new schema, you basically have to use it throughout the entire provider and controller or else everything will be broken. Maybe this is a good way to split up the code changes into more reasonable changes. If we go with that approach, the PR's will be:

1. Add new fields to the interface
2. This PR, which does preflight checking for both the old CloudstackDatacenterConfig schema and the new schema, but uses old CloudstackDatacenterConfig attributes when generating the CAPI/CAPC templates
3. Third PR, which will invoke the SetDefaults method any time we're deserializing a CloudstackDatacenterConfig and then exclusively use the new fields through the provider and controller, but generating CAPI/CAPC templates for v0.4.5
4. Fourth PR, which will support CAPC v0.5 for template generation

Does that sound reasonable?

Reply:
Sounds good to me.